File: classes-ObjC/STCardsMenu/STCardsMenuChildVC.m
//
//  STCardsMenuChildVC.m
//  STCardsMenu-ObjC
//
//  Created by TangJR on 12/14/15.
//  Copyright © 2015 tangjr. All rights reserved.
//

#import "STCardsMenuChildVC.h"

@interface STCardsMenuChildVC () {
    UIColor *_tintColor;
}

@end

@implementation STCardsMenuChildVC

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    self.hamburgButton.alpha = 0;
}

#pragma mark - Button Tapped

- (void)hamburgButtonTapped:(STCardsMenuHamburgButton *)sender {
    if (self.delegate && [self.delegate respondsToSelector:@selector(cardsMenu:hamburgButtonTapped:)]) {
        [self.delegate cardsMenu:self hamburgButtonTapped:sender];
    }
}

#pragma mark - Getter / Setter

- (void)setTitle:(NSString *)title {
    [super setTitle:title];
    self.titleLabel.text = title;
}

- (void)setTintColor:(UIColor *)tintColor {
    _tintColor = tintColor;
    self.hamburgButton.tintColor = tintColor;
    self.titleLabel.textColor = tintColor;
}

- (UIColor *)tintColor {
    if (_tintColor) {
        _tintColor = [UIColor whiteColor];
    }
    return _tintColor;
}

- (STCardsMenuHamburgButton *)hamburgButton {
    if (!_hamburgButton) {
        _hamburgButton = [STCardsMenuHamburgButton new];
        _hamburgButton.tintColor = self.tintColor;
        _hamburgButton.translatesAutoresizingMaskIntoConstraints = NO;
        [_hamburgButton addTarget:self action:@selector(hamburgButtonTapped:) forControlEvents:UIControlEventTouchUpInside];
        [self.view addSubview:_hamburgButton];
        NSLayoutConstraint *leftConstraint = [NSLayoutConstraint constraintWithItem:_hamburgButton attribute:NSLayoutAttributeLeft relatedBy:NSLayoutRelationEqual toItem:self.view attribute:NSLayoutAttributeLeft multiplier:1 constant:10];
        NSLayoutConstraint *topConstraint =  [NSLayoutConstraint constraintWithItem:_hamburgButton attribute:NSLayoutAttributeTop relatedBy:NSLayoutRelationEqual toItem:self.view attribute:NSLayoutAttributeTop multiplier:1 constant:10];
        NSLayoutConstraint *widthConstraint = [NSLayoutConstraint constraintWithItem:_hamburgButton attribute:NSLayoutAttributeWidth relatedBy:NSLayoutRelationEqual toItem:nil attribute:NSLayoutAttributeNotAnAttribute multiplier:1 constant:40];
        NSLayoutConstraint *heightConstraint = [NSLayoutConstraint constraintWithItem:_hamburgButton attribute:NSLayoutAttributeHeight relatedBy:NSLayoutRelationEqual toItem:nil attribute:NSLayoutAttributeNotAnAttribute multiplier:1 constant:30];
        [self.view addConstraints:@[leftConstraint, topConstraint]];
        [_hamburgButton addConstraints:@[widthConstraint, heightConstraint]];
    }
    return _hamburgButton;
}

- (UILabel *)titleLabel {
    if (!_titleLabel) {
        _titleLabel = [UILabel new];
        _titleLabel.textColor = [UIColor whiteColor];
        _titleLabel.translatesAutoresizingMaskIntoConstraints = NO;
        [self.view addSubview:_titleLabel];
        NSLayoutConstraint *leftConstraint = [NSLayoutConstraint constraintWithItem:_titleLabel attribute:NSLayoutAttributeLeft relatedBy:NSLayoutRelationEqual toItem:self.view attribute:NSLayoutAttributeLeft multiplier:1 constant:20];
        NSLayoutConstraint *topConstraint =  [NSLayoutConstraint constraintWithItem:_titleLabel attribute:NSLayoutAttributeTop relatedBy:NSLayoutRelationEqual toItem:self.view attribute:NSLayoutAttributeTop multiplier:1 constant:10];
        NSLayoutConstraint *heightConstraint = [NSLayoutConstraint constraintWithItem:_titleLabel attribute:NSLayoutAttributeHeight relatedBy:NSLayoutRelationEqual toItem:nil attribute:NSLayoutAttributeNotAnAttribute multiplier:1 constant:30];
        [self.view addConstraints:@[leftConstraint, topConstraint]];
        [_titleLabel addConstraints:@[heightConstraint]];
    }
    return _titleLabel;
}

@end

File: classes-Swift/STCardsMenu/STCardsMenuChildVC.swift
//
//  STCardsMenuChildVC.swift
//  STCardsMenu-Swift
//
//  Created by TangJR on 12/15/15.
//  Copyright © 2015 tangjr. All rights reserved.
//

import UIKit

protocol STCardsMenuChildVCDelegate: class {
    
    func hamburgButtonTapped(childVC: STCardsMenuChildVC, hamburgButton: STCardsMenuHamburgButton)
}

class STCardsMenuChildVC: UIViewController {
    
    var tintColor: UIColor?
    weak var delegate: STCardsMenuChildVCDelegate?
    let hamburgButton = STCardsMenuHamburgButton()
    let titleLabel = UILabel()

    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view.
    }
    
    override func viewWillAppear(animated: Bool) {
        super.viewWillAppear(animated)
        setupHamburgButton()
        setupTitlabel()
    }
}

extension STCardsMenuChildVC {
    private func setupHamburgButton() {
        view.addSubview(hamburgButton)
        hamburgButton.hidden = true
        hamburgButton.tintColor = tintColor
        hamburgButton.addTarget(self, action: Selector("hamburgButtonTapped:"), forControlEvents: .TouchUpInside)
        hamburgButton.translatesAutoresizingMaskIntoConstraints = false
        let leftConstraint = NSLayoutConstraint(item: hamburgButton, attribute: .Left, relatedBy: .Equal, toItem: view, attribute: .Left, multiplier: 1.0, constant: 20.0)
        let topConstriant = NSLayoutConstraint(item: hamburgButton, attribute: .Top, relatedBy: .Equal, toItem: view, attribute: .Top, multiplier: 1.0, constant: 20.0)
        let widthConstraint = NSLayoutConstraint(item: hamburgButton, attribute: .Width, relatedBy: .Equal, toItem: nil, attribute: .NotAnAttribute, multiplier: 1.0, constant: 40.0)
        let heightConstraint =  NSLayoutConstraint(item: hamburgButton, attribute: .Height, relatedBy: .Equal, toItem: nil, attribute: .NotAnAttribute, multiplier: 1.0, constant: 30.0)
        hamburgButton.addConstraints([widthConstraint, heightConstraint])
        view.addConstraints([leftConstraint, topConstriant])
    }
    
    private func setupTitlabel() {
        view.addSubview(titleLabel)
        titleLabel.text = title
        titleLabel.textColor = tintColor
        titleLabel.translatesAutoresizingMaskIntoConstraints = false
        let leftConstraint = NSLayoutConstraint(item: titleLabel, attribute: .Left, relatedBy: .Equal, toItem: view, attribute: .Left, multiplier: 1.0, constant: 10.0)
        let topConstraint = NSLayoutConstraint(item: titleLabel, attribute: .Top, relatedBy: .Equal, toItem: view, attribute: .Top, multiplier: 1.0, constant: 10.0)
        let heightConstraint = NSLayoutConstraint(item: titleLabel, attribute: .Height, relatedBy: .Equal, toItem: nil, attribute: .NotAnAttribute, multiplier: 1.0, constant: 30.0)
        titleLabel.addConstraints([heightConstraint])
        view.addConstraints([leftConstraint, topConstraint])
    }
}

extension STCardsMenuChildVC {
    func hamburgButtonTapped(sender: STCardsMenuHamburgButton) {
        delegate?.hamburgButtonTapped(self, hamburgButton: sender)
    }
}